src/coprocessor/dag/executor/aggregation.rs

```diff
@@ -159,35 +157,41 @@ impl<'a> AggregationExecutor<'a> {
     }
 }
 
-impl<'a> Executor for AggregationExecutor<'a> {
+impl Executor for AggregationExecutor {
     fn next(&mut self) -> Result<Option<Row>> {
         if !self.executed {
             self.aggregate()?;
             self.executed = true;
         }
 
-        if self.cursor >= self.group_keys.len() {
-            return Ok(None);
-        }
-        // calc all aggr func
         let mut aggr_cols = Vec::with_capacity(2 * self.aggr_func.len());
-        let group_key = &self.group_keys[self.cursor];
-        let mut aggrs = self.group_key_aggrs.remove(group_key).unwrap();
-        for aggr in &mut aggrs {
-            aggr.calc(&mut aggr_cols)?;
-        }
-        // construct row data
-        let value_size = group_key.len() + approximate_size(&aggr_cols, false);
-        let mut value = Vec::with_capacity(value_size);
-        box_try!(value.encode(aggr_cols.as_slice(), false));
-        if !self.group_by.is_empty() {
-            value.extend_from_slice(group_key);
+        match self.group_key_aggrs.get_index_mut(self.cursor) {
+            Some((ref group_key, ref mut aggrs)) => {
+                self.cursor += 1;
+
+                // calc all aggr func
+                for aggr in aggrs.iter_mut() {
+                    aggr.calc(&mut aggr_cols)?;
+                }
+
+                // construct row data
+                let value_size = group_key.len() + approximate_size(&aggr_cols, false);
+                let mut value = Vec::with_capacity(value_size);
+                box_try!(value.encode(aggr_cols.as_slice(), false));
+                if !self.group_by.is_empty() {
+                    value.extend_from_slice(group_key);
+                }
+                Ok(Some(Row {
+                    handle: 0,
```

Comment:
I recall that handle is deprecated, it can be removed later.

Reply:
How about do it in next PR? now there are still several places use `Row.handle`.